feat(collab): Add user presence notifications and cross-device support

charlenetcy opened the pull request:
## Collaborative Editing: User Presence Notifications and Cross-Device Support

Adds real-time user presence notifications and enables cross-device collaborative editing.

### Changes
- Toast notifications for user join and leave events
- Cross-device collaboration via configurable host URL (ngrok support)
- Generate temporary random usernames (adjective + animal) as placeholder until user service integration
- Fix username displaying as default value, "Coding buddy", when first user leaves the room
- Supabase environment validation with clear error messages
- Improved color palette for better cursor and username visibility

### Technical Changes
- Add `VITE_NGROK_COLLAB_HOST` env variable to enable cross-device collaborative testing until PartyKit deployment
- Refactor awareness event handler for cleaner event processing and proper TypeScript types 
- Populate username map during initial sync to track all users correctly
- Resolve ghost user accumulation after component unmount

Temporary random usernames will be replaced with actual user data once user service integration is complete.

---

## How to Activate Cross-Device Collaboration

### Person 1 (Host)
Sign up on [ngrok](https://ngrok.com/) and get token from https://dashboard.ngrok.com
```bash
brew install ngrok

# One-time setup
ngrok config add-authtoken YOUR_TOKEN 

# Start services
docker-compose up

# In new terminal, create tunnel
ngrok http 8082
# Copy the domain (e.g., abc123.ngrok-free.app)
```

### Person 2 (Partner)
```bash
# Paste the URL EXCLUDING THE "HTTPS:// " from Person 1 in the .env  
VITE_NGROK_COLLAB_HOST=abc123.ngrok-free.app

# Start frontend only (excluding collaboration-service)
docker-compose up frontend 
```

Both navigate to `http://localhost:3000/room/` and input same room id to test real-time collaboration.

**Note:** Only Person 1 runs `collaboration-service`. Person 2 connects to it via ngrok tunnel.

charlenetcy committed:
feat(collaboration): Add user presence toast notifications

Display real-time notifications when users join or leave collaborative editing rooms.

- Add toast notifications for user join or leave events using react-hot-toast
- Generate temporary random usernames (adjective + animal) as placeholder until user service integration
- Implement awareness event handling with duplicate prevention and initial sync filtering
- Add Supabase environment variable validation in server
- Remove StrictMode to fix WebSocket double-mounting in development

Random names serve as placeholder for actual usernames from user service.

charlenetcy committed:
feat(collaboration): Allow cross-device collaborative editing

-Refactored the collaboration provider host configuration with an environment variable lookup, facilitating the testing of the collaboration service's core functionality.

- Enhanced the randomised username generation by updating `adjectives` with developer-themed terms for a more fitting and engaging user experience.

The use of tunneling services like ngrok, acts as a temporary solution until the dedicated PartyKit deployment is established closer to production.

charlenetcy committed:
fix(collaboration): Resolve username display when leaving bug

- Fix first user's username not displaying when leaving collaboration room bug
- Refactor awareness event handler

charlenetcy committed:
Merge remote-tracking branch 'origin/develop' into feature/collaborative-editing

charlenetcy committed:
fix(collab): Resolve ghost user accumulation after component unmount

leepoeaik reviewed (approved):
LGTM!

charlenetcy committed:
fix(collab): env issue

ChinZJ commented on frontend/src/collaboration/hooks/useCollabEditor.ts:
```diff
@@ -2,43 +2,178 @@
+    });
+    
+    // Clears debounce flag after timeout to allow future notifications
+    setTimeout(() => {
```
Minor nitpick
Not too familiar with this but can I check if this is what might happen?

1. User A creates the room
2. User B joins
3. User B disconnects
    1. User B gets added to this debounce which is 15s
4. User A leaves
    1. No one in the room, cleanup occurs (line166)
5. If timeout has not fired, it will result in memory leak?

Given frequency of room joining I do not think this is an issue, but might be worth handling

charlenetcy replied:
I see what you mean, nice catch! 
Will implement the fix!

ChinZJ reviewed (commented):
Unsure about one part, else LGTM!

charlenetcy committed:
Merge branch 'develop' into feature/collaborative-editing

charlenetcy committed:
fix: Prevent memory leaks in user disconnect handling

Address PR feedback regarding potential memory leaks with debounced
user disconnect notifications.
Implement proper timeout reference tracking and cleanup to ensure all resources are properly released when component
unmounts.
